### PR TITLE
Improve code coverage of System.Linq.Expressions.IndexExpression

### PR DESCRIPTION
**Pull request by DenisIstomin**
Improve code coverage of System.Linq.Expressions.IndexExpression to 100%; small Improvements of System.Linq.Expressions.ExpressionVisitor code coverage.

Contributes to #1198

сс @VSadov @stephentoub

### CODE REVIEW

**Comment by DenisIstomin** [comment]
@joshfree, @VSadov, could you please take a look?

File: src/System.Linq.Expressions/src/Properties/InternalsVisibleTo.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+//------------------------------------------------------------
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("System.Linq.Expressions.Tests, PublicKey=002400000480000094000000060200000024000052534131000400000100010015c01ae1f50e8cc09ba9eac9147cf8fd9fce2cfe9f8dce4f7301c4132ca9fb50ce8cbf1df4dc18dd4d210e4345c744ecb3365ed327efdbc52603faa5e21daa11234c8c4a73e51f03bf192544581ebe107adee3a34928e39d04e524a9ce729d5090bfd7dad9d10c722c0def9ccc08ff0a03790e48bcd1f9b6c476063e1966a1c4")]
```
Comment (stephentoub) [inline src/System.Linq.Expressions/src/Properties/InternalsVisibleTo.cs]: I believe we're trying to avoid using InternalsVisibleTo.  @weshaggard?

Comment (DenisIstomin) [inline src/System.Linq.Expressions/src/Properties/InternalsVisibleTo.cs]: There are 5 projects that using `InternalsVisibleTo` to expose "internals" to Tests project:
1. System.Xml.XDocument
2. System.Reflection.Metadata
3. System.Diagnostics.Debug
4. System.Collections.Immutable
5. System.Text.Encodings.Web

That's why I applied that attribute to `System.Linq.Expressions`. Do I need to remove it?

Comment (weshaggard) [inline src/System.Linq.Expressions/src/Properties/InternalsVisibleTo.cs]: I didn't look deeply into the usage here but yes our general principle is to avoid InternalsVisibleTo because we want to run these same tests on the other frameworks (i.e. .NET Framework, .NET Native) and by taking internal dependencies these tests will not work.

Comment (JonHanna) [inline src/System.Linq.Expressions/src/Properties/InternalsVisibleTo.cs]: It also makes it harder to spot dead code if that code is perhaps being hit by a test that internals are visible to, but can't be hit by any real code.

Comment (VSadov) [inline src/System.Linq.Expressions/src/Properties/InternalsVisibleTo.cs]: Right. It would be definitely preferable to not add more IVTs 
IVT is a tool of last resort. They do add to the maintenance burden and cause various portability problems.

File: src/System.Linq.Expressions/src/System.Linq.Expressions.csproj [hunk 2]
```diff
@@ -229,4 +229,4 @@
     <TargetingPackReference Include="System.Core" />
   </ItemGroup>
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />
-</Project>
+</Project>
```
Comment (stephentoub) [inline src/System.Linq.Expressions/src/System.Linq.Expressions.csproj]: Did this file change at all other than whitespace?  Looks like it didn't, in which case can you just revert this change?

File: src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    public static class IndexExpressionHelpers
+    {
+        internal static void AssertEqual(Expressions.IndexExpression expected, Expressions.IndexExpression actual)
+        {
+            Assert.Equal(expected.Object, actual.Object);
+            Assert.Equal(expected.Indexer, actual.Indexer);
+            Assert.Equal(expected.Arguments, actual.Arguments);
+        }
+
+        internal static void AssertInvokeCorrect<T>(T expected, Expressions.IndexExpression expr, SampleClassWithProperties parameter)
```
Comment (stephentoub) [inline src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionHelpers.cs]: Is the parameter argument unused?

File: src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionTests.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+{
+    public static class IndexExpressionTests
+    {
+        [Fact]
+        public static void UpdateSameTest()
+        {
+            var instance = new SampleClassWithProperties { DefaultProperty = new List<int> { 100, 101 } };
+            Expressions.IndexExpression expr = instance.DefaultIndexExpression;
+
+            var exprUpdated = expr.Update(expr.Object, expr.Arguments);
```
Comment (stephentoub) [inline src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionTests.cs]: Nit: please use the type name rather than var

File: src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionTests.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        public static void UpdateTest()
+        {
+            var instance = new SampleClassWithProperties
+            {
+                DefaultProperty = new List<int> { 100, 101 },
+                AlternativeProperty = new List<int> { 200, 201 }
+            };
+
+            var expr = instance.DefaultIndexExpression;
+            var newProperty = Expression.Property(Expression.Constant(instance),
```
Comment (stephentoub) [inline src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionTests.cs]: I'll stop commenting on the vars, but in general, please use the type name unless the type is explicitly mentioned on the right-hand side, e.g. a cast, a ctor, etc.

File: src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionVisitorTests.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                    {expr.Object, newObject}
+                };
+
+                for (int i = 0; i < expr.Arguments.Count; i++)
+                {
+                    _dict.Add(expr.Arguments[i], newArguments?[i]);
+                }
+            }
+
+            private readonly Dictionary<Expression, Expression> _dict;
```
Comment (stephentoub) [inline src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionVisitorTests.cs]: Please move the fields up to before the ctors

File: src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionVisitorTests.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+            public override Expression Visit(Expression node)
+            {
+                return _dict.ContainsKey(node)
+                    ? _dict[node]
+                    : base.Visit(node);
```
Comment (stephentoub) [inline src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionVisitorTests.cs]: Nit: perf doesn't matter so much in these tests, but this would be better as:
```C#
Expression result;
return _dict.TryGetValue(node, out result) ? result : base.Visit(node);
```
so as to avoid the second lookup.

File: src/System.Linq.Expressions/tests/IndexExpression/SampleClassWithProperties.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
+﻿using System.Collections.Generic;
+using System.Reflection;
+
+namespace System.Linq.Expressions.Tests.IndexExpression
+{
+    internal class SampleClassWithProperties
+    {
+        internal readonly string DefaultPropertyName = "DefaultProperty";
+        internal readonly string AlternativePropertyName = "AlternativeProperty";
```
Comment (stephentoub) [inline src/System.Linq.Expressions/tests/IndexExpression/SampleClassWithProperties.cs]: Can these be const rather than readonly?

**Comment by stephentoub** [comment]
cc: @JonHanna, @bartdesmet

File: src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+            Assert.Equal(expected.Arguments, actual.Arguments);
+        }
+
+        internal static void AssertInvokeCorrect<T>(T expected, Expressions.IndexExpression expression)
+        {
+            var lambda = Expression.Lambda<Func<T>>(expression);
+
+            // Compile and evaluate with interpretation flag and without
+            // in case there are bugs in the compiler/interpreter. 
+            Assert.Equal(expected, lambda.Compile(false).Invoke());
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionHelpers.cs]: Nit: can these use delegate invocation syntax?

File: src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
+﻿using Xunit;
+
+namespace System.Linq.Expressions.Tests.IndexExpression
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionHelpers.cs]: For consistency with other tests, this could just be `System.Linq.Expressions.Tests`. It also removes the need for qualification or aliasing of `IndexExpression`.

File: src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionHelpers.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
+﻿using Xunit;
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionHelpers.cs]: Will need copyright header in all these files.

**Comment by stephentoub** [comment]
Test Innerloop CentOS7.1 Release Build and Test please

**Comment by stephentoub** [comment]
Thanks, @DenisIstomin. LGTM
